docs/source/wrappers.rst

```diff
@@ -0,0 +1,487 @@
+
+Compilation of ``Model``
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+You have two options to compile your model:
+1. Compile your model within ``model_build_fn`` and return this
+compiled model. In this case, SciKeras will not re-compile your model
+and all compilation parameters (such as ``optimizer``) given to
+:py:func:`scikeras.wrappers.BaseWrapper.__init__` will be ignored.
+2. Return an uncompiled model from ``model_build_fn`` and let
+SciKeras handle the compilation. In this case, SciKeras will
+apply all of the compilation parameters, including instantiating
+losses, metrics and optimizers.
+
+The first route will be more flexible if you wish to determine how to compile
+your ``Model`` within the same function in which you define it. The latter will
+offer an easy way to compile and tune compilation parameters. Examples:
+
+.. code:: python
+
+    def model_build_fn(compile_kwargs):
+        # you can access the ``optimizer`` param here
+        optimizer = compile_kwargs["optimizer"]
+        if optimizer is None:
+            # and apply any custom logic you wish
+            ...
+        model = Model(...)
+        ...
+        model.compile(optimizer=optimizer)
+        return model
+
+    clf = KerasClassifier(model=model_build_fn)
+    clf.fit(X, y)
+    y_pred = clf.predict(X_valid)
+
+.. code:: python
+
+    from tensorflow.keras.optimizers import Adam
+
+    def model_build_fn():
+        model = Model(...)
+        ...
+        # Do not call model.compile
+        return model  # That's it, SciKeras will compile your model for you
+
+    clf = KerasClassifier(model=model_build_fn, optimizer=Adam)
+    clf.fit(X, y)
+    y_pred = clf.predict(X_valid)
+
+
+Arguments to ``model``/``model_build_fn``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+You probably wish to pass parameters from :py:class:`~scikeras.wrappers.BaseWrapper`
+to ``model``, or you may want to use attributes from
+:py:class:`~scikeras.wrappers.BaseWrapper` such as ``n_features_in_`` while building
+your model. SciKeras allows you to do both.
+
+To enable this, SciKeras uses two special arguments to ``model`` that will only
+be passed if they are present in ``model``'s signature (i.e. there is an argument
+with the same name in ``model``'s signature):
+
+``meta_params``
++++++++++++++++
+This is a dictionary containing all of the attributes that
+:py:class:`~scikeras.wrappers.BaseWrapper` creates when it is initialized
+These include ``n_features_in_``, ``y_dtype_``, etc. For a full list,
+see the :ref:`scikeras-api` documentation.
+
+``compile_kwargs``
+++++++++++++++++++
+This is a dictionary of parameters destined for :py:func:`tensorflow.Keras.Model.compile`.
+You will want to accept this parameter unless you are returning an un-compiled ``Model``
+instance. Parameters available via this dictionary are:
+
+* ``optimizer``
+* ``loss``
+* ``callbacks``
+* Any other parameters with the prefix ``optimizer__``, ``loss__`` or ``callbacks__``
```

Comment:
Right, assuming that the user passed either an optimizer name or class to `optimizer`. Obviously if they did not provied an optimizer but passed optimizer parameters, that's a valueerror

Reply:
How would a user tune `optimizer__momentum`? Shouldn't `compile_kwargs` contain legal keyword arguments to `model.compile`?

Reply:
> How would a user tune optimizer__momentum?

If the user compiles themselves:
```python3
def build_fn(compile_kwargs):
     momentum=compile_kwargs["optimizer__momentum"]
    ...
    model.compile(optimizer=SGD(momentum=momentum))
    return model
```

Letting SciKeras manage compilation:
```python3
KerasClassifier(..., optimizer=SGD, optimizer__momentum=0.5)
```

Reply:
Why not pass the instantiated loss/optimizer `compile_kwargs`? They'll have to be created anyways if the user returns an uncompiled model.

``` python
def build_fn(compile_kwargs=None):
    assert isinstance(compile_kwargs["optimizer], tf.keras.optimizer.Optimizer)  # or whatever class for Keras Optimizers
    ...
```

Then, the user won't have to write any more code to tune `optimizer__momentum`, and `compile_kwargs` will always be valid arguments for `model.compile`.

It looks like this would require #66. In that framework, the implementation would look something like this:

``` python
compile_kwargs = {k: _get_compile_attr(k) for k in ["loss", "optimizer", "metrics", "run_eagerly"]}
model = self._final_build_fn(compile_kwargs=compile_kwargs)
if not _not_compiled(model):
    self._compile_keras_model(**compile_kwargs)
```

Reply:
I'm not opposed to _adding_ compiled optimizers into `compile_kwargs` but I am against _only_ including the compiled optimizers. I would like the users to have access to the raw parameters they inputted so that they can do with them as they wish.

Reply:
>  I would like the users to have access to the raw parameters they inputted so that they can do with them as they wish

Then let's fix that problem. How 'bout this signature?

``` python
def model_build_fn(compile_kwargs=None, params=None, meta=None):
    """
    compile_kwargs : dict
        Valid keyword arguments for model.compile. This includes Keras Optimizers and Losses in the
        ``loss`` and ``optimizer`` keys.
    params : dict
        The raw parameters of the ``BaseWrapper`` class, and the output of ``BaseWrapper.get_params``.
        These are the parameters that have been provided by the user.
    meta : dict
        Meta information about the number of features, number of expected outputs, etc.
    """
```

Reply:
I like that!

And we'll also be checking the signature and providing any parameters that are named the same as `__init__` parameters, right?

Reply:
Again thank you for the suggestion. This is now implemented in #67 . I'll leave this open here since I do need to update these docs to match.